### PR TITLE
Add descriptor_tests covering tr(), and fix minor bugs

### PR DESCRIPTION
**Pull request by sipa**
This fixes two bugs in the current logic for `tr()` descriptors:
* ToPrivateString does not always work, because the provided private key may mismatch the parity of the x-only public key.
* The descriptors inferred for `pk()` inside `tr()` have the wrong x-only flag, leading to such descriptors generating the wrong scriptPubKey (roundtripping through ToString does fix it however, so this seems unobservable in the current code).

These were discovered while adding unit tests to descriptor_tests that cover various aspects of `tr()` descriptors, which are now also added here.

### CODE REVIEW

File: src/script/descriptor.cpp [hunk 1]
```diff
@@ -259,7 +259,15 @@ class ConstPubkeyProvider final : public PubkeyProvider
...
-        if (!arg.GetKey(m_pubkey.GetID(), key)) return false;
+        if (m_xonly) {
+            for (const auto& keyid : XOnlyPubKey(m_pubkey).GetKeyIDs()) {
+                arg.GetKey(keyid, key);
+                if (key.IsValid()) break;
+            }
+        } else {
+            arg.GetKey(m_pubkey.GetID(), key);
+        }
+        if (!key.IsValid()) return false;
```
[the remark this answers is not in the document]
Comment (sipa) [reply src/script/descriptor.cpp:270]: Done. Also updated PR description.

**Comment by achow101** [comment]
ACK 9f5c7830ee36e5a9fa9a8a816316fe01b62eb0de

**Comment by sipa** [comment]
Improved the new tests a bit, adding a case with multiple tree branches, and signing with mixed pub/priv descriptors.

**Comment by laanwj** [comment]
Added to 23.0 milestone because bugfix.

**Comment by achow101** [comment]
ACK 0683f377e1588758da86368f82efee765f89d890

**Review by instagibbs — APPROVED** [review]
ACK https://github.com/bitcoin/bitcoin/pull/24343/commits/0683f377e1588758da86368f82efee765f89d890

File: src/script/descriptor.cpp [hunk 2]
```diff
@@ -1253,7 +1261,7 @@ std::unique_ptr<DescriptorImpl> InferScript(const CScript& script, ParseScriptCo
 {
     if (ctx == ParseScriptContext::P2TR && script.size() == 34 && script[0] == 32 && script[33] == OP_CHECKSIG) {
         XOnlyPubKey key{Span{script}.subspan(1, 32)};
-        return std::make_unique<PKDescriptor>(InferXOnlyPubkey(key, ctx, provider));
+        return std::make_unique<PKDescriptor>(InferXOnlyPubkey(key, ctx, provider), true);
```
Comment (jonatack) [inline src/script/descriptor.cpp:1264]: 18ad54c3
```suggestion
        return std::make_unique<PKDescriptor>(InferXOnlyPubkey(key, ctx, provider), /*xonly=*/true);
```
Verified `descriptor_tests` fail without this commit:
`test/descriptor_tests.cpp(327): error: in "descriptor_tests/descriptor_test": check spks_inferred[0] == spks[n] has failed`

File: src/test/descriptor_tests.cpp [hunk 3]
```diff
@@ -269,7 +309,12 @@ void DoCheck(const std::string& prv, const std::string& pub, const std::string&
                     CMutableTransaction spend;
                     spend.vin.resize(1);
                     spend.vout.resize(1);
-                    BOOST_CHECK_MESSAGE(SignSignature(Merge(keys_priv, script_provider), spks[n], spend, 0, 1, SIGHASH_ALL), prv);
+                    std::vector<CTxOut> utxos(1);
+                    PrecomputedTransactionData txdata;
+                    txdata.Init(spend, std::move(utxos), /* force */ true);
+                    MutableTransactionSignatureCreator creator(&spend, 0, CAmount{0}, &txdata, SIGHASH_DEFAULT);
```
Comment (jonatack) [inline src/test/descriptor_tests.cpp:315]: 0683f37
```diff
-                    txdata.Init(spend, std::move(utxos), /* force */ true);
-                    MutableTransactionSignatureCreator creator(&spend, 0, CAmount{0}, &txdata, SIGHASH_DEFAULT);
+                    txdata.Init(spend, std::move(utxos), /*force=*/true);
+                    MutableTransactionSignatureCreator creator(&spend, /*input_idx=*/0, CAmount{0}, &txdata, SIGHASH_DEFAULT);
```

**Review by jonatack — COMMENTED** [review]
Code review ACK 0683f377e1588758da86368f82efee765f89d890

A couple of nitty suggestions in the comments, feel free to ignore.

**Comment by michaelfolkson** [comment]
Are you going to open a PR with your nits @jonatack? Maybe should be included in 23.0 milestone if you are?

**Comment by jonatack** [comment]
@michaelfolkson no, if the author considers it worthwhile they could be done in a follow-up.

**Comment by michaelfolkson** [comment]
@jonatack: Ok will wait to see whether @sipa thinks the nits are worth it or not.

This was merged before #24433 but since that was merged the guidance is:

"You are expected to reply to any review comments before your pull request is merged. You may update the code or reject the feedback if you do not agree with it, but you should express so in a reply." :)

**Comment by jonatack** [comment]
It's more likely I'd follow up for, say, a bug or missing tests.

**Pull request merged**